internal/controller/clustertunnel_controller.go

```diff
@@ -145,9 +147,25 @@ func (r *ClusterTunnelReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	if err := r.Get(ctx, req.NamespacedName, tunnel); err != nil {
 		if apierrors.IsNotFound(err) {
 			// Tunnel object not found, could have been deleted after reconcile request.
-			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
-			// Return and don't requeue
-			r.log.Info("Tunnel deleted, nothing to do")
+			// Owned objects are automatically garbage collected.
+			objectClient, err := k8s.NewObjectClient(r.Client, &r.log)
+			if err != nil {
+				return ctrl.Result{}, err
```

Comment:
This logic is in the wrong place (ref: the error you saw). When this condition is hit, the tunnel is already removed from etcd. 

What you need is, instead of the logic in both here and `tunnel_controller.go`, move this to https://github.com/adyanth/cloudflare-operator/blob/d1570e5b098d688258e1b60f6fac44e25357d2d8/internal/controller/generic_tunnel_reconciler.go#L157-L162

Reply:
Oh, derp. Thanks will fix this

Reply:
Did you get around to this? Would love to merge this in

Reply:
All good. Life comes first!

Reply:
Just wanted to give you another heads up, might be another week